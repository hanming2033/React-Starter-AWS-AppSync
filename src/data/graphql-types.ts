/* tslint:disable */
//  This file was automatically generated and should not be edited.

export interface GetLocalStatesQuery {
  forms:  {
    __typename: "forms",
    input_Email: string,
  } | null,
  nav:  {
    __typename: "nav",
    nextPath: string,
  } | null,
};
