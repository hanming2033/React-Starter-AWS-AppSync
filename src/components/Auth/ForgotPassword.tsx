import * as React from 'react'
import { Form, Field, Formik, FormikActions, FormikProps } from 'formik'
import { Query, QueryResult } from 'react-apollo'
import { GetLocalStatesQuery } from '../../data/graphql-types'
import { GET_LOCAL_STATES } from '../../data/actions/Queries'
import * as yup from 'yup'
import { TChangeComponent } from './AuthenticatorRouter'
import { AuthProxy } from './AuthProxies/AuthProxy'

// *1 define the form values interface
interface IRequestFormValues {
  email: string
}
interface IResetFormValues {
  email: string
  code: string
  password: string
  confirmPassword: string
}

export interface IForgotPasswordProps {
  toComp: TChangeComponent
}

export interface IForgotPasswordState {}

// *2 define yup schema for form validation
const schemaRequest = yup.object().shape({
  email: yup
    .string()
    .email('Not a valid email')
    .required('Email is required')
})
const schemaReset = yup.object().shape({
  email: yup
    .string()
    .email('Not a valid email')
    .required('Email is required'),
  code: yup.string().required('Authentication code is required'),
  password: yup.string().required('Password is required'),
  confirmPassword: yup
    .string()
    .oneOf([yup.ref('password'), null], 'Passwords must match')
    .required('Password confirm is required')
})

// *3 create actual form outsourced from Formik
const formRequest = ({ errors, touched, isSubmitting }: FormikProps<IRequestFormValues>) => (
  <Form>
    <Field name="email" placeholder="Email" />
    {touched.email && errors.email}
    <br />
    <button disabled={isSubmitting}>Send Code</button>
  </Form>
)
const formReset = ({ errors, touched, isSubmitting }: FormikProps<IResetFormValues>) => (
  <Form>
    <Field name="email" placeholder="Email" disabled />
    {touched.email && errors.email}
    <br />
    <Field name="code" placeholder="Auth Code" />
    {touched.code && errors.code}
    <br />
    <Field type="password" name="password" placeholder="Password" />
    {touched.password && errors.password}
    <br />
    <Field type="password" name="confirmPassword" placeholder="Password" />
    {touched.confirmPassword && errors.confirmPassword}
    <br />
    <button disabled={isSubmitting}>Reset Password</button>
  </Form>
)

class ForgotPassword extends React.Component<IForgotPasswordProps, IForgotPasswordState> {
  public state = { delivery: null }

  // *4 create onsubmit method
  public requestCode = async (
    values: IRequestFormValues,
    formikBag: FormikActions<IRequestFormValues>,
    qryRes: QueryResult<GetLocalStatesQuery>
  ) => {
    formikBag.setSubmitting(true)
    // store email in local state
    if (!qryRes.data || !qryRes.data.forms) return
    const newData = {
      ...qryRes.data,
      forms: {
        ...qryRes.data.forms,
        input_Email: values.email
      }
    }
    qryRes.client.writeData({ data: newData })

    const res = await AuthProxy.requestForgotPasswordCode(values.email)
    if (res.data) {
      this.setState({ delivery: res.data.CodeDeliveryDetails })
      formikBag.resetForm()
      formikBag.setSubmitting(false)
    } else if (res.error) {
      formikBag.setErrors({
        email: (res.error.message as string).includes('Username/client') ? 'Email Not Found' : ''
      })
      formikBag.setFieldValue('password', '', false)
      formikBag.setSubmitting(false)
      if (res.error && (res.error.message as string).includes('no registered/verified')) {
        this.props.toComp('confirmSignUp') // *good
      }
    }
  }

  public resetPassword = async (values: IResetFormValues, formikBag: FormikActions<IResetFormValues>) => {
    formikBag.setSubmitting(true)
    const res = await AuthProxy.resetPassword(values.email, values.code, values.password)
    if (res.data) {
      formikBag.resetForm()
      formikBag.setSubmitting(false)
      this.setState({ delivery: null })
      this.props.toComp('signIn') // *good
    } else if (res.error) {
      formikBag.setErrors({
        code: res.error.code ? res.error.message : '',
        password: (res.error.message as string).includes('password') ? res.error.message : '',
        confirmPassword: (res.error.message as string).includes('password') ? res.error.message : ''
      })
      formikBag.setFieldValue('password', '', false)
      formikBag.setSubmitting(false)
    }
  }

  public render() {
    return (
      <Query<GetLocalStatesQuery> query={GET_LOCAL_STATES} fetchPolicy="no-cache">
        {qryRes => {
          if (qryRes.error) return <h1>Error!!</h1>
          if (!qryRes.data || !qryRes.data.forms) return null

          return (
            <>
              {/* 5inject Formik component into view */}
              <h1>Forgot Password</h1>
              <Formik
                // get current user email from global state first
                initialValues={{
                  email: qryRes.data.forms.input_Email || '',
                  code: '',
                  password: '',
                  confirmPassword: ''
                }}
                validationSchema={this.state.delivery ? schemaReset : schemaRequest}
                onSubmit={
                  this.state.delivery
                    ? (values: any, formikBag) => this.resetPassword(values, formikBag)
                    : (values, formikBag) => this.requestCode(values, formikBag, qryRes)
                }
                component={this.state.delivery ? formReset : formRequest}
              />
              <button onClick={() => this.props.toComp('signIn')}>Back to Sign In</button>
            </>
          )
        }}
      </Query>
    )
  }
}

export default ForgotPassword
