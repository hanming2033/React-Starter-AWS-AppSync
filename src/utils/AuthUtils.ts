import { QueryResult } from 'react-apollo'
import { GetLocalStatesQuery } from '../data/graphql-types'

export const updateCacheForm = (qryRes: QueryResult<GetLocalStatesQuery>, propName: string, value: string) => {
  if (qryRes.data && qryRes.data.forms) {
    const newData: GetLocalStatesQuery = {
      ...qryRes.data,
      forms: {
        ...qryRes.data.forms,
        [propName]: value
      }
    }
    qryRes.client.writeData({ data: newData })
  }
}
