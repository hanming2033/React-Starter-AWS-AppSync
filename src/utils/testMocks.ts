import { GetLocalStatesQuery } from '../data/graphql-types'
import { QueryResult } from 'react-apollo'
import { ApolloError } from 'apollo-client'
import { client } from '../AWSApolloClient'

export const getLocalStateData: GetLocalStatesQuery = {
  auth: {
    __typename: 'auth',
    isAuthenticated: false
  },
  forms: {
    __typename: 'forms',
    input_Email: ''
  },
  nav: {
    __typename: 'nav',
    nextPath: ''
  }
}

export const qryRes: QueryResult<GetLocalStatesQuery> = {
  client,
  data: undefined,
  error: undefined,
  loading: false,
  networkStatus: 7,
  startPolling: jest.fn(),
  stopPolling: jest.fn(),
  subscribeToMore: jest.fn(),
  variables: {},
  refetch: jest.fn(),
  fetchMore: jest.fn(),
  updateQuery: jest.fn()
}

export const apolloError: ApolloError = {
  name: 'error',
  message: 'error',
  graphQLErrors: [{ name: 'error', message: 'error' }],
  networkError: null,
  extraInfo: ''
}
