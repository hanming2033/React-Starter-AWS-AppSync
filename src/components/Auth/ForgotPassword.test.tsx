import * as React from 'react'
import * as enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import enzymeSerializer from 'enzyme-to-json/serializer'
import ForgotPassword, { FormikRequestCode } from './ForgotPassword'
import { Query, QueryResult } from 'react-apollo'
import { qryRes, apolloError, getLocalStateData } from '../../utils/testMocks'
import { GetLocalStatesQuery } from '../../data/graphql-types'
import wait from 'waait'

expect.addSnapshotSerializer(enzymeSerializer)
enzyme.configure({ adapter: new Adapter() })

describe('<ForgotPassword /> Main Suite', () => {
  it('snapshot on loading state', () => {
    const queryWrapper = enzyme.shallow(<ForgotPassword toComp={jest.fn()} />)
    const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, loading: true }
    const wrapper = enzyme.shallow(queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
    expect(wrapper).toMatchSnapshot()
  })
  it('snapshot on error state', () => {
    const queryWrapper = enzyme.shallow(<ForgotPassword toComp={jest.fn()} />)
    const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, error: apolloError }
    const wrapper = enzyme.shallow(queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
    expect(wrapper).toMatchSnapshot()
  })
  it('snapshot on default state : request code', () => {
    const queryWrapper = enzyme.shallow(<ForgotPassword toComp={jest.fn()} />)
    const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
    const wrapper = enzyme.shallow(<div>{queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>}</div>)
    expect(wrapper).toMatchSnapshot()
  })
  it('snapshot on default state : reset code', async () => {
    const queryWrapper = enzyme.shallow(<ForgotPassword toComp={jest.fn()} />)
    const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
    queryWrapper.setState({ delivery: { error: 'error' } })
    const wrapper = enzyme.shallow(<div>{queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>}</div>)
    queryWrapper.update()
    expect(wrapper).toMatchSnapshot()
  })
  describe('Back to Sign In Button Suite', () => {
    describe('Interactions', () => {
      // how interaction on this feature affects itself, other component, state(rendering in children), prop method call
      it('should call toComp with args signIn', () => {
        const toComp = jest.fn()
        const queryWrapper = enzyme.shallow(<ForgotPassword toComp={toComp} />)
        const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
        const wrapper = enzyme.shallow(<div>{queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>}</div>)
        wrapper.find('button').simulate('click')
        expect(toComp).toBeCalledWith('signIn')
      })
    })
  })
})

describe('<FormikRequestCode /> Main Suite', () => {
  // rendering based on different state and props of main component
  // test here is state and props does not directly affects children
  // test here is state and props affects more than 1 children
  it('snapshot on default state', () => {
    const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
    const wrapper = enzyme.mount(FormikRequestCode(newRes, jest.fn(), jest.fn()))
    expect(wrapper).toMatchSnapshot()
  })
  describe('Email Input Suite', () => {
    describe('Static/Conditional Rendering', () => {
      it('snapshot on qryRes has email:"a@b.c" as email input', () => {
        const newRes: QueryResult<GetLocalStatesQuery> = {
          ...qryRes,
          data: { ...getLocalStateData, forms: { __typename: 'forms', input_Email: 'a@b.c' } }
        }
        const wrapper = enzyme.mount(FormikRequestCode(newRes, jest.fn(), jest.fn()))
        expect(wrapper.find('input[name="email"]')).toMatchSnapshot()
      })
    })
    describe('Interactions', () => {
      // how interaction on this feature affects itself, other component, state(rendering in children), prop method call
      it('should render "Email is required" error on blur ', async () => {
        const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
        const wrapper = enzyme.mount(FormikRequestCode(newRes, jest.fn(), jest.fn()))
        wrapper.find('input[name="email"]').simulate('change', { persist: () => undefined, target: { name: 'email', value: '' } })
        wrapper.find('input[name="email"]').simulate('blur', { persist: () => undefined, target: { name: 'email' } })
        await wait(0)
        wrapper.update()
        expect(wrapper.find('[children="Email is required"]').length).toBe(1)
      })
      it('should render "hello" on simulate change', async () => {
        const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
        const wrapper = enzyme.mount(FormikRequestCode(newRes, jest.fn(), jest.fn()))
        wrapper.find('input[name="email"]').simulate('change', { persist: () => undefined, target: { name: 'email', value: 'hello' } })
        wrapper.find('input[name="email"]').simulate('blur', { persist: () => undefined, target: { name: 'email' } })
        await wait(0)
        wrapper.update()
        expect(wrapper.find('input[name="email"]').props().value).toBe('hello')
      })
      it('should render "Not a valid email" error on invalid email & blur', async () => {
        const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
        const wrapper = enzyme.mount(FormikRequestCode(newRes, jest.fn(), jest.fn()))
        wrapper.find('input[name="email"]').simulate('change', { persist: () => undefined, target: { name: 'email', value: 'hello' } })
        wrapper.find('input[name="email"]').simulate('blur', { persist: () => undefined, target: { name: 'email' } })
        await wait(0)
        wrapper.update()
        expect(wrapper.find('[children="Not a valid email"]').length).toBe(1)
      })
      it('should render no error on valid email & blur', async () => {
        const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
        const wrapper = enzyme.mount(FormikRequestCode(newRes, jest.fn(), jest.fn()))
        wrapper
          .find('input[name="email"]')
          .simulate('change', { persist: () => undefined, target: { name: 'email', value: 'hello@hi.com' } })
        wrapper.find('input[name="email"]').simulate('blur', { persist: () => undefined, target: { name: 'email' } })
        await wait(0)
        wrapper.update()
        expect(wrapper.find('[children="Not a valid email"]').length).toBe(0)
        expect(wrapper.find('[children="Email is required"]').length).toBe(0)
      })
    })
  })
})
