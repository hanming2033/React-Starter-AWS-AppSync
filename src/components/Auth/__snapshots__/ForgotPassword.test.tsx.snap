// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgotPassword /> Main Suite snapshot on default state : request code 1`] = `
<div>
  <React.Fragment>
    <h1>
      Forgot Password
    </h1>
    <Formik
      enableReinitialize={false}
      initialValues={
        Object {
          "email": "",
        }
      }
      isInitialValid={false}
      onSubmit={[Function]}
      render={[Function]}
      validateOnBlur={true}
      validateOnChange={true}
      validationSchema={
        ObjectSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_defaultDefault": [Function],
          "_deps": Array [],
          "_excludedEdges": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_nodes": Array [
            "email",
          ],
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "object",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "fields": Object {
            "email": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": false,
                "undefined": false,
              },
              "_mutate": false,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
            },
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        }
      }
    />
    <button
      onClick={[Function]}
    >
      Back to Sign In
    </button>
  </React.Fragment>
</div>
`;

exports[`<ForgotPassword /> Main Suite snapshot on default state : reset code 1`] = `
<div>
  <React.Fragment>
    <h1>
      Forgot Password
    </h1>
    <Formik
      enableReinitialize={false}
      initialValues={
        Object {
          "code": "",
          "confirmPassword": "",
          "email": "",
          "password": "",
        }
      }
      isInitialValid={false}
      onSubmit={[Function]}
      render={[Function]}
      validateOnBlur={true}
      validateOnChange={true}
      validationSchema={
        ObjectSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_defaultDefault": [Function],
          "_deps": Array [],
          "_excludedEdges": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_nodes": Array [
            "confirmPassword",
            "password",
            "code",
            "email",
          ],
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "object",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "fields": Object {
            "code": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": false,
              },
              "_mutate": false,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
            },
            "confirmPassword": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": false,
              },
              "_mutate": false,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {
                  null,
                },
                "refs": Map {
                  "password" => Reference {
                    "_get": [Function],
                    "isContext": false,
                    "isSelf": false,
                    "key": "password",
                    "map": [Function],
                    "path": "password",
                    "prefix": "$",
                  },
                },
              },
              "_whitelistError": [Function],
              "tests": Array [
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
            },
            "email": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": false,
                "undefined": false,
              },
              "_mutate": false,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
            },
            "password": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": false,
              },
              "_mutate": false,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
            },
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        }
      }
    />
    <button
      onClick={[Function]}
    >
      Back to Sign In
    </button>
  </React.Fragment>
</div>
`;

exports[`<ForgotPassword /> Main Suite snapshot on error state 1`] = `
<h1>
  Error!!
</h1>
`;

exports[`<ForgotPassword /> Main Suite snapshot on loading state 1`] = `
<h1>
  Loading...
</h1>
`;

exports[`<FormikRequestCode /> Main Suite Email Input Suite Static/Conditional Rendering snapshot on qryRes has email:"a@b.c" as email input 1`] = `
<input
  name="email"
  onBlur={[Function]}
  onChange={[Function]}
  placeholder="Email"
  value="a@b.c"
/>
`;

exports[`<FormikRequestCode /> Main Suite snapshot on default state 1`] = `
<Formik
  enableReinitialize={false}
  initialValues={
    Object {
      "email": "",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  render={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": false,
      "_nodes": Array [
        "email",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "fields": Object {
        "email": StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": false,
            "undefined": false,
          },
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
>
  <Form>
    <Component
      formik={
        Object {
          "dirty": false,
          "errors": Object {},
          "handleBlur": [Function],
          "handleChange": [Function],
          "handleReset": [Function],
          "handleSubmit": [Function],
          "initialValues": Object {
            "email": "",
          },
          "isSubmitting": false,
          "isValid": false,
          "isValidating": false,
          "registerField": [Function],
          "resetForm": [Function],
          "setError": [Function],
          "setErrors": [Function],
          "setFieldError": [Function],
          "setFieldTouched": [Function],
          "setFieldValue": [Function],
          "setFormikState": [Function],
          "setStatus": [Function],
          "setSubmitting": [Function],
          "setTouched": [Function],
          "setValues": [Function],
          "submitCount": 0,
          "submitForm": [Function],
          "touched": Object {},
          "unregisterField": [Function],
          "validate": undefined,
          "validateField": [Function],
          "validateForm": [Function],
          "validateOnBlur": true,
          "validateOnChange": true,
          "validationSchema": ObjectSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_defaultDefault": [Function],
            "_deps": Array [],
            "_excludedEdges": Array [],
            "_exclusive": Object {},
            "_mutate": false,
            "_nodes": Array [
              "email",
            ],
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "object",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "fields": Object {
              "email": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": false,
                  "undefined": false,
                },
                "_mutate": false,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                  [Function],
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
            },
            "tests": Array [],
            "transforms": Array [
              [Function],
            ],
          },
          "values": Object {
            "email": "",
          },
        }
      }
    >
      <form
        onSubmit={[Function]}
      >
        <C
          name="email"
          placeholder="Email"
        >
          <FieldInner
            formik={
              Object {
                "dirty": false,
                "errors": Object {},
                "handleBlur": [Function],
                "handleChange": [Function],
                "handleReset": [Function],
                "handleSubmit": [Function],
                "initialValues": Object {
                  "email": "",
                },
                "isSubmitting": false,
                "isValid": false,
                "isValidating": false,
                "registerField": [Function],
                "resetForm": [Function],
                "setError": [Function],
                "setErrors": [Function],
                "setFieldError": [Function],
                "setFieldTouched": [Function],
                "setFieldValue": [Function],
                "setFormikState": [Function],
                "setStatus": [Function],
                "setSubmitting": [Function],
                "setTouched": [Function],
                "setValues": [Function],
                "submitCount": 0,
                "submitForm": [Function],
                "touched": Object {},
                "unregisterField": [Function],
                "validate": undefined,
                "validateField": [Function],
                "validateForm": [Function],
                "validateOnBlur": true,
                "validateOnChange": true,
                "validationSchema": ObjectSchema {
                  "_blacklist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "_conditions": Array [],
                  "_defaultDefault": [Function],
                  "_deps": Array [],
                  "_excludedEdges": Array [],
                  "_exclusive": Object {},
                  "_mutate": false,
                  "_nodes": Array [
                    "email",
                  ],
                  "_options": Object {
                    "abortEarly": true,
                    "recursive": true,
                  },
                  "_type": "object",
                  "_typeError": [Function],
                  "_whitelist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "fields": Object {
                    "email": StringSchema {
                      "_blacklist": RefSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_conditions": Array [],
                      "_deps": Array [],
                      "_exclusive": Object {
                        "required": false,
                        "undefined": false,
                      },
                      "_mutate": false,
                      "_options": Object {
                        "abortEarly": true,
                        "recursive": true,
                      },
                      "_type": "string",
                      "_typeError": [Function],
                      "_whitelist": RefSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "tests": Array [
                        [Function],
                        [Function],
                        [Function],
                      ],
                      "transforms": Array [
                        [Function],
                      ],
                    },
                  },
                  "tests": Array [],
                  "transforms": Array [
                    [Function],
                  ],
                },
                "values": Object {
                  "email": "",
                },
              }
            }
            name="email"
            placeholder="Email"
          >
            <input
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Email"
              value=""
            />
          </FieldInner>
        </C>
        <span />
        <br />
        <button
          disabled={false}
        >
          Send Code
        </button>
      </form>
    </Component>
  </Form>
</Formik>
`;
